# David-young

[**YOUNG**]{.ul}

-   The first cell and gene quality control is in [QC_young](src/QC_01.Rmd)

-   The normalisation by deconvolution is in [normalise_young](src/normalise_01.Rmd)

-   Feature selection and dimensional reduction in [feature_selection_dimred_young](src/feature_selection_dimred_01.Rmd)

-   Batch correction in [batchcorrect_young](src/batchcorrect.Rmd)

-   Clustering at different resolutions, [clustering_01](src/clustering_01.Rmd)

-  First rough annotation in [annotation_01](src/annotation_01.Rmd) 

- [ClusterQC](src/cluster_QC_01.Rmd) 

-  Stricter cell and gene QC; dimensional reduction and feature selection with the cleaned data[QC_norm_featrueselect_dimred_02](src/cluster_QC_01.Rmd/QC_norm_featrueselect_dimred_02_jpriller)

-  Clustering with different resolutions [clustering_02](src/clustering_02.Rmd_stop)

-  Annotation [annotation_02](src/annotation_02.Rmd_stop) - the marker plot is missing

-  Differential gene expression (DE) between WT and KO and between WT and HETs for each cluster [DE_WT_KO_HET_clusters](src/DE_WT_KO_HET_clusters.Rmd),  [DE_WT_HET_microglia](src/DE_WT_HET_microglia_edgeR) and [DE_WT_HET_BAMs](src/DE_WT_HET_BAMs_edgeR)

 -   Interactive [Shinyapp](https://annawilliams.shinyapps.io/shinyApp_jpriller) with the analysis until this stage.
 
-   Differential abundance of cells between clusters between WT KO and HET  [DA_miloR](src/DA_miloR.Rmd)
 
**To Do/ In Progress**
 
- DE using MAST? 

- Add tissue-genotype label in Shiny

- Trajectory analysis

PS: links might not work