---
title: "Differential Expression"
author: "Nadine Bestard"
date: "21/07/2021"
output:
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### set-up
```{r}
library(scran) # for scDE
library(scater) # for aggregate counts
library(edgeR) #for De
library(here) # reproducible paths
```
```{r load}
source(here("src/colours.R"))
project <- "fire-mice"
sce <- readRDS(here("processed", project, "sce_nodbl_anno_02.RDS")) 
clusters_name <- "clusters_named"
```


## Use of pseudo-bulk samples

We perform the DE analysis separately for each label to identify cell type-specific transcriptional effects of KO condition. The actual DE testing is performed on “pseudo-bulk” expression profiles (Tung et al. 2017), generated by summing counts together for all cells with the same combination of label and sample. This leverages the resolution offered by single-cell technologies to define the labels, and combines it with the statistical rigor of existing methods for DE analyses involving a small number of samples.

```{r sum}
# each column represents unique combination for each cluster and each sample replicate
summed <- aggregateAcrossCells(sce, 
    id=colData(sce)[,c(clusters_name, "Sample")]) 
summed
```

## Perform DE

### Filter replicates that do not have enough cells

We first filter out all the combinations sample-label that resulted in less than 10
cells. These are saved in /outs/`r project`/DE_cluster/fail_min_cells_DE.csv

```{r filter-summed}
# Removing all pseudo-bulk samples with 'insufficient' cells.
summed_filt <- summed[, summed$ncells >= 10]

# and saving to file which one there are
fail_n_cells  <- colData(summed)[ colData(summed)$ncells < 10, c("Sample", "mouse", "genotype", "tissue", "chip", clusters_name, "ncells")]

write.csv(fail_n_cells, here("outs", project,"DE_edgeR", "fail_min_cells_DE.csv"), row.names = FALSE)
```

### Design 

`tissue` and `chip` are confounded, meaning it is not possible to correct for both; and seen there is greater variation in `chip` than `tissue` correcting for `chip` seems more sensible.

```{r}
table(sce$tissue, sce$chip)
table(sce$genotype, sce$chip)
table(sce$genotype, sce$tissue)
design=~factor(chip) + genotype
```

We then compute the DE with edgeR, one of the best methods according to [(Soneson et al. 2018)](https://www.nature.com/articles/nmeth.4612). The results are saved in /outs/`r project`/DE_edgeR/de_results

**Output:**

LogFC is the log fold-change, which is the log difference between both groups

LogCPM are the log counts per million, which can be understood as measuring expression level.

F is the F-statistic from the quasi-likelihood F-test.

PValue is the nominal p-value derived from F without any multiple testing correction

FDR (False discovery rate) is the PValue after Benjamini-Hochberg correction.  For example, the set of genes with adjusted p value less than 0.1 should contain no more than 10% false positives.

If the comparison was not possible for a particular cluster, most commonly due to lack of residual degrees of freedom ( e.g. an absence of enough samples from both conditions) it is listed in /outs/`r project`/DE_edgeR/fail_min_cells_DE.csv

```{r de, eval=FALSE}
# Reordering the genotype factor, treated should be second level
summed_filt$genotype <- factor(summed_filt$genotype, levels = c("WT", "HET", "KO"))

# KO ones
# run de
de_results_KO <- pseudoBulkDGE(summed_filt, 
    label=summed_filt$clusters_named,
    design= design,
    coef="genotypeKO",
    condition=summed_filt$genotype 
)

# HET ones
de_results_HET <- pseudoBulkDGE(summed_filt, 
    label=summed_filt$clusters_named,
    design= design,
    coef="genotypeHET",
    condition=summed_filt$genotype 
)
# save result
saveRDS(de_results_HET, here("processed",project, "DE_results_edgeR_HET.RDS"))
saveRDS(de_results_KO, here("processed",project, "DE_results_edgeR_KO.RDS"))
lapply(names(de_results_KO), function(x){
  de_results_clu <- de_results_KO[[x]]
  write.csv(de_results_clu, here("outs", project, "DE_edgeR", "de_results_KO", paste0("de_results_KO_", x, ".csv")), quote = FALSE)
}
)
lapply(names(de_results_HET), function(x){
  de_results_clu <- de_results_HET[[x]]
  write.csv(de_results_clu, here("outs", project, "DE_edgeR", "de_results_HET", paste0("de_results_HET_", x, ".csv")), quote = FALSE)
}
)

# save fail
write.csv(metadata(de_results_HET), here("outs", project,"DE_edgeR", "fail_contrast_DE_HET.csv"), row.names = FALSE)

# save fail
write.csv(metadata(de_results_KO), here("outs", project,"DE_edgeR", "fail_contrast_DE_KO.csv"), row.names = FALSE)

```

## Filter the results

For each one of the labels a list is produced with the DE genes at a FDR of 5%. Summaries of these results are saved in /outs/`r project`/DE_edgeR/ 

Values of 1, -1 and 0 indicate that the gene is significantly upregulated, downregulated or not significant, respectively.
Genes listed as NA were either filtered out as low-abundance genes for a given label’s analysis, or the comparison of interest was not possible for that particular label.

de_filtered_genecounts.csv contains a small summary with the number of genes DE for each one of the clusters. 

de_filtered_genes.csv contains the list of genes DE. 

## Plot the results

Saved in /outs/`r project`/DE_edgeR/de_filtered_plots

```{r filter-de, eval=FALSE}
de_results <- list(KO=de_results_KO, HET=de_results_HET)
cols <- list(KO=col_wt_het_ko[3], HET=col_wt_het_ko[2], WT=col_wt_het_ko[1])
for(gnt in c("KO", "HET")){
  
de_results <- de_results[[gnt]]
is_de <- decideTestsPerLabel(de_results, threshold=0.05)
summary_de <- summarizeTestsPerLabel(is_de)

# virtually 0 or NA we do not care, we want only + or -
is_de_0 <- is_de
is_de_0[is.na(is_de_0)]<- 0
# but filter the original object to get them back
is_de_flt <- is_de[(rowSums(abs(is_de_0)) != 0), ]

write.csv(summary_de, here("outs", project,"DE_edgeR", paste0( "de_", gnt, "_filtered_genecounts.csv")))

write.csv(is_de_flt, here("outs", project, "DE_edgeR", paste0( "de_", gnt, "_filtered_genes.csv" )))

is_de_genes <- rownames(is_de_flt)
for( gene in is_de_genes){
pdf(file = here("outs", project, "DE_edgeR", "plots", paste0( "de_", gnt, "_filtered_plots", gene, ".pdf")))
plot <- plotExpression(logNormCounts(sce),
    features=gene,
    x="genotype", colour_by="genotype",
    other_fields="clusters_named") +
    facet_wrap(~clusters_named) +
    scale_color_manual(values = col_wt_het_ko) +
    ggtitle(gene)
print(plot)
dev.off()
}

}
```